Sources/SwiftInspectorCore/TypeLocationAnalyzer.swift

```diff
@@ -0,0 +1,174 @@
+
+    switch node {
+    case let classDecl as ClassDeclSyntax:
+      anyLocatedType = (classDecl.identifier.text, classDecl.classKeyword, classDecl.modifiers)
+    case let enumDecl as EnumDeclSyntax:
+      anyLocatedType = (enumDecl.identifier.text, enumDecl.enumKeyword, enumDecl.modifiers)
+    case let protocolDecl as ProtocolDeclSyntax:
+      anyLocatedType = (protocolDecl.identifier.text, protocolDecl.protocolKeyword, protocolDecl.modifiers)
+    case let structDecl as StructDeclSyntax:
+      anyLocatedType = (structDecl.identifier.text, structDecl.structKeyword, structDecl.modifiers)
+    default:
+      anyLocatedType = nil
+    }
+
+    if let anyLocatedType = anyLocatedType {
+      var indexOfStartingLine = currentLineNumber
+      // We need to add this in early. We don't modify currentLineNumber since they will be added
+      // in later when we compute the leading newlines for this entire node.
+      indexOfStartingLine += countOfLeadingNewlinesForType(
+        keywordToken: anyLocatedType.keywordToken,
+        modifiers: anyLocatedType.modifiers)
+
+      let indexOfEndingLine = indexOfStartingLine + countOfNewlines(within: node)
+
+      let locatedType = LocatedType(
+        name: anyLocatedType.name,
+        indexOfStartingLine: indexOfStartingLine,
+        indexOfEndingLine: indexOfEndingLine)
+      onNodeVisit(locatedType)
+    }
+
+    if let leadingTrivia = node.leadingTrivia {
+      currentLineNumber += countOfNewlines(from: leadingTrivia, for: node)
+    }
+    if let trailingTrivia = node.trailingTrivia {
+      currentLineNumber += countOfNewlines(from: trailingTrivia, for: node)
+    }
+
+    return super.visitAny(node)
+  }
+
+  var currentLineNumber: UInt = 0
+  let onNodeVisit: (LocatedType) -> Void
+
+  /// The total newlines associated with this node.
+  private func countOfNewlines(from trivia: Trivia, for node: Syntax) -> UInt {
+    // Some nodes seem to include trivia from other nodes. Filtering to just tokens ensures we get
+    // an accurate count.
+    guard node is TokenSyntax else { return 0 }
+    return trivia.countOfNewlines()
+  }
+
+  /// The number of newlines preceding this token.
+  private func countOfLeadingNewlinesForType(
+    keywordToken: TokenSyntax,
+    modifiers: ModifierListSyntax?) -> UInt
+  {
+    var result: UInt = 0
+    result += keywordToken.leadingTrivia.countOfNewlines()
+    modifiers?.leadingTrivia.flatMap { result += $0.countOfNewlines() }
+    return result
+  }
+
+  /// Find the number of newlines within this node.
+  private func countOfNewlines(within node: Syntax) -> UInt {
+    var countOfNewlinesInsideType: UInt = 0
+
+    for (offset, token) in node.tokens.enumerated() {
+      // We've already counted the leading trivia for the first token.
+      if let leadingTrivia = token.leadingTrivia, offset != 0 {
+        countOfNewlinesInsideType += leadingTrivia.countOfNewlines()
+      }
+      if let trailingTrivia = token.trailingTrivia{
```

Comment:
I wondered if we needed to not count to trailing trivia for the last token, but it seems to work 🤷‍♂